### PR TITLE
CAN parser dict

### PR DESCRIPTION
**Pull request by robbederks**
Gets rid of the hard-coded CAN bus names (pt, body, cam, chassis, adas, ...) in favor of a dict

### CODE REVIEW

**Comment by sshane** [comment]
Why do we need this?

@deanlee and I were thinking about one CANParser where bus is a key in the message tuple to reduce copies

**Comment by robbederks** [comment]
For Tesla the predefined bus names don't make much sense (and potentially for other cars too?), so this is mainly  a cleanup for that.

I'd also prefer a single CANParser, but that sounds like a lot more effort?

**Review by adeebshihadeh — APPROVED** [review]
Feels like a lateral move to me, but feel free to merge if you feel strongly

File: opendbc/car/body/carstate.py [hunk 1]
```diff
@@ -30,10 +31,12 @@ def update(self, cp, *_) -> structs.CarState:
...
+  def get_can_parsers(CP):
     messages = [
       ("MOTORS_DATA", 100),
       ("VAR_VALUES", 10),
       ("BODY_DATA", 1),
     ]
-    return CANParser(DBC[CP.carFingerprint]["pt"], messages, 0)
+    return {
+      'main': CANParser(DBC[CP.carFingerprint]["main"], messages, 0)
+    }
```
Comment (sshane) [inline opendbc/car/body/carstate.py:42]: I don't generally like furthering the string dictionary keys for this stuff, an enum sounds much better.

Comment (robbederks) [reply opendbc/car/body/carstate.py:42]: guess we can define an enum with some standard bus names somewhere

File: opendbc/car/body/carcontroller.py [hunk 2]
```diff
@@ -15,9 +15,9 @@
 
 
 class CarController(CarControllerBase):
-  def __init__(self, dbc_name, CP):
-    super().__init__(dbc_name, CP)
-    self.packer = CANPacker(dbc_name)
+  def __init__(self, dbc_names, CP):
+    super().__init__(dbc_names, CP)
+    self.packer = CANPacker(dbc_names[Bus.MAIN])
```
Comment (sshane) [inline opendbc/car/body/carcontroller.py:20]:
```suggestion
    self.packer = CANPacker(dbc_names[Bus.main])
```

File: opendbc/car/body/carstate.py [hunk 3]
```diff
@@ -30,10 +31,12 @@ def update(self, cp, *_) -> structs.CarState:
...
+  def get_can_parsers(CP):
     messages = [
       ("MOTORS_DATA", 100),
       ("VAR_VALUES", 10),
       ("BODY_DATA", 1),
     ]
-    return CANParser(DBC[CP.carFingerprint]["pt"], messages, 0)
+    return {
+      Bus.MAIN: CANParser(DBC[CP.carFingerprint][Bus.MAIN], messages, 0)
+    }
```
Comment (sshane) [inline opendbc/car/body/carstate.py:42]:
```suggestion
    return {Bus.MAIN: CANParser(DBC[CP.carFingerprint][Bus.MAIN], messages, 0)}
```

this is fine, and GH shows the diff better

File: opendbc/car/body/values.py [hunk 4]
```diff
@@ -7,7 +7,6 @@
 
 SPEED_FROM_RPM = 0.008587
 
-
 class CarControllerParams:
```
Comment (sshane) [inline opendbc/car/body/values.py:10]: two spaces before/after class is the standard Python stylization, and functional PRs should mostly not touch style for diff reasons

File: opendbc/car/ford/carstate.py [hunk 5]
```diff
@@ -168,9 +167,13 @@ def get_cam_can_parser(CP):
...
         ("Side_Detect_L_Stat", 5),
         ("Side_Detect_R_Stat", 5),
       ]
 
-    return CANParser(DBC[CP.carFingerprint]["pt"], messages, CanBus(CP).camera)
+    return {
+      Bus.PT: CANParser(DBC[CP.carFingerprint][Bus.PT], pt_messages, CanBus(CP).main),
+      Bus.CAM: CANParser(DBC[CP.carFingerprint][Bus.PT], cam_messages, CanBus(CP).camera),
+    }
+
```
Comment (sshane) [inline opendbc/car/ford/carstate.py:179]: extra line

File: opendbc/car/ford/radar_interface.py [hunk 6]
```diff
@@ -95,13 +95,14 @@ def __init__(self, CP):
...
+    if CP.radarUnavailable:
+      return
 
     self.updated_messages = set()
     self.track_id = 0
-    self.radar = DBC[CP.carFingerprint]['radar']
-    if CP.radarUnavailable:
-      self.rcp = None
-    elif self.radar == RADAR.DELPHI_ESR:
+    self.radar = DBC[CP.carFingerprint][Bus.RADAR]
```
Comment (sshane) [inline opendbc/car/ford/radar_interface.py:104]:
```suggestion
    self.radar = DBC[CP.carFingerprint].get(Bus.RADAR)
```

can also just do this

File: opendbc/car/tesla/carstate.py [hunk 7]
```diff
@@ -96,14 +98,14 @@ def get_can_parser(CP):
...
       ("DAS_status", 2),
       ("SCCM_steeringAngleSensor", 100),
     ]
 
-    return CANParser(DBC[CP.carFingerprint]['pt'], messages, CANBUS.autopilot_party)
+    return {
+      Bus.PARTY: CANParser(DBC[CP.carFingerprint][Bus.PARTY], party_messages, CANBUS.party),
+      Bus.AP_PARTY: CANParser(DBC[CP.carFingerprint][Bus.PARTY], ap_party_messages, CANBUS.autopilot_party)
+    }
+
```
Comment (sshane) [inline opendbc/car/tesla/carstate.py:111]: extra line

File: opendbc/car/toyota/carstate.py [hunk 8]
```diff
@@ -219,63 +222,62 @@ def get_can_parser(CP):
...
+        cam_messages += [
           ("PRE_COLLISION", 33),
         ]
 
-    return CANParser(DBC[CP.carFingerprint]["pt"], messages, 2)
+    return {
+      Bus.PT: CANParser(DBC[CP.carFingerprint][Bus.PT], pt_messages, 0),
+      Bus.CAM: CANParser(DBC[CP.carFingerprint][Bus.PT], cam_messages, 2),
+    }
+
```
Comment (sshane) [inline opendbc/car/toyota/carstate.py:283]: extra line

File: opendbc/car/toyota/values.py [hunk 9]
```diff
@@ -73,16 +73,18 @@ class ToyotaCarDocs(CarDocs):
   package: str = "All"
   car_parts: CarParts = field(default_factory=CarParts.common([CarHarness.toyota_a]))
 
+def toyota_dbc_dict(pt, radar):
```
Comment (sshane) [inline opendbc/car/toyota/values.py:76]:
```suggestion
def dbc_dict(pt, radar):
```

File: opendbc/car/tests/test_platform_configs.py [hunk 10]
```diff
@@ -9,7 +9,7 @@ def test_configs(self, subtests):
         assert platform.config._frozen
 
         if platform != "MOCK":
-          assert "pt" in platform.config.dbc_dict
+          assert len(platform.config.dbc_dict.keys()) > 0
```
Comment (sshane) [inline opendbc/car/tests/test_platform_configs.py:12]:
```suggestion
          assert len(platform.config.dbc_dict) > 0
```

File: opendbc/car/gm/values.py [hunk 11]
```diff
@@ -195,7 +199,6 @@ class CanBus:
   LOOPBACK = 128
   DROPPED = 192
```
Comment (sshane) [inline opendbc/car/gm/values.py:201]:
```suggestion


```

File: opendbc/car/honda/carstate.py [hunk 12]
```diff
@@ -270,38 +274,34 @@ def update(self, cp, cp_cam, _, cp_body, __) -> structs.CarState:
...
-    return CANParser(DBC[CP.carFingerprint]["pt"], messages, CanBus(CP).camera)
+    body_messages = [
+      ("BSM_STATUS_LEFT", 3),
+      ("BSM_STATUS_RIGHT", 3),
+    ]
+
+    return {
+      Bus.PT: CANParser(DBC[CP.carFingerprint][Bus.PT], pt_messages, CanBus(CP).pt),
+      Bus.CAM: CANParser(DBC[CP.carFingerprint][Bus.PT], cam_messages, CanBus(CP).camera),
+      Bus.BODY: CANParser(DBC[CP.carFingerprint][Bus.BODY], body_messages, CanBus(CP).radar) if CP.enableBsm else None,
```
Comment (sshane) [inline opendbc/car/honda/carstate.py:305]: it simplifies things if we ensure every key has a value

File: opendbc/car/honda/values.py [hunk 13]
```diff
@@ -116,6 +116,9 @@ def init(self):
     self.flags |= HondaFlags.NIDEC
 
 
+def honda_with_radar_dbc_dict(pt_dict):
+  return {Bus.PT: pt_dict, Bus.RADAR: 'acura_ilx_2016_nidec'}
+
```
Comment (sshane) [inline opendbc/car/honda/values.py:121]:
```suggestion


```

File: opendbc/car/hyundai/carstate.py [hunk 14]
```diff
@@ -263,11 +269,64 @@ def update_canfd(self, cp, cp_cam) -> structs.CarState:
 
     return ret
 
-  def get_can_parser(self, CP):
+
```
Comment (sshane) [inline opendbc/car/hyundai/carstate.py:272]:
```suggestion
```

only one line needed for methods

File: opendbc/car/hyundai/carstate.py [hunk 15]
```diff
@@ -263,11 +269,64 @@ def update_canfd(self, cp, cp_cam) -> structs.CarState:
...
+      cam_messages += [
+        ("SCC_CONTROL", 50),
+      ]
+
+    return {
+      Bus.PT: CANParser(DBC[CP.carFingerprint][Bus.PT], pt_messages, CanBus(CP).ECAN),
+      Bus.CAM: CANParser(DBC[CP.carFingerprint][Bus.PT], cam_messages, CanBus(CP).CAM),
+    }
+
+
```
Comment (sshane) [inline opendbc/car/hyundai/carstate.py:324]:
```suggestion
```

**Review by sshane — CHANGES_REQUESTED** [review]

File: opendbc/car/honda/carstate.py (hunk 12 above)
Comment (robbederks) [reply opendbc/car/honda/carstate.py:305]: agree, this is why I had the `None` check.
guess I could rewrite as a conditional sum of dicts?

File: opendbc/car/interfaces.py [hunk 16]
```diff
@@ -221,19 +217,19 @@ def configure_torque_tune(candidate: str, tune: structs.CarParams.LateralTuning,
...
       if cp is not None:
         cp.update_strings(can_packets)
 
     # get CarState
     ret = self._update()
 
-    ret.canValid = all(cp.can_valid for cp in self.can_parsers if cp is not None)
-    ret.canTimeout = any(cp.bus_timeout for cp in self.can_parsers if cp is not None)
+    ret.canValid = all(cp.can_valid for cp in self.can_parsers.values() if cp is not None)
+    ret.canTimeout = any(cp.bus_timeout for cp in self.can_parsers.values() if cp is not None)
```
Comment (sshane) [inline opendbc/car/interfaces.py:232]:
```suggestion
    ret.canValid = all(cp.can_valid for cp in self.can_parsers.values())
    ret.canTimeout = any(cp.bus_timeout for cp in self.can_parsers.values())
```

File: opendbc/car/__init__.py [hunk 17]
```diff
@@ -84,11 +84,20 @@ def scale_tire_stiffness(mass, wheelbase, center_to_front, tire_stiffness_factor
   return tire_stiffness_front, tire_stiffness_rear
 
 
-DbcDict = dict[str, str]
+DbcDict = dict[str | StrEnum, str]
```
Comment (sshane) [inline opendbc/car/__init__.py:87]: it should always be StrEnum now right?

File: opendbc/car/honda/carstate.py [hunk 18]
```diff
@@ -299,9 +300,11 @@ def get_can_parsers(self, CP):
...
     ]
 
-    return {
+    parsers = {
       Bus.pt: CANParser(DBC[CP.carFingerprint][Bus.pt], pt_messages, CanBus(CP).pt),
       Bus.cam: CANParser(DBC[CP.carFingerprint][Bus.pt], cam_messages, CanBus(CP).camera),
-      Bus.body: CANParser(DBC[CP.carFingerprint][Bus.body], body_messages, CanBus(CP).radar) if CP.enableBsm else None,
     }
+    if CP.enableBsm:
+      parsers += {Bus.body: CANParser(DBC[CP.carFingerprint][Bus.body], body_messages, CanBus(CP).radar)}
```
Comment (sshane) [inline opendbc/car/honda/carstate.py:308]: dict + dict  isn't a valid operation, try `|=`

File: opendbc/car/honda/carstate.py [hunk 19]
```diff
@@ -270,38 +275,36 @@ def update(self, cp, cp_cam, _, cp_body, __) -> structs.CarState:
...
+    }
     if CP.enableBsm:
-      messages = [
-        ("BSM_STATUS_LEFT", 3),
-        ("BSM_STATUS_RIGHT", 3),
-      ]
-      bus_body = CanBus(CP).radar # B-CAN is forwarded to ACC-CAN radar side (CAN 0 on fake ethernet port)
-      return CANParser(DBC[CP.carFingerprint]["body"], messages, bus_body)
-    return None
+      parsers += {Bus.body: CANParser(DBC[CP.carFingerprint][Bus.body], body_messages, CanBus(CP).radar)}
```
Comment (sshane) [inline opendbc/car/honda/carstate.py:308]:
```suggestion
      parsers[Bus.body] = CANParser(DBC[CP.carFingerprint][Bus.body], body_messages, CanBus(CP).radar)
```

dict + dict isn't valid

Comment (robbederks) [reply opendbc/car/honda/carstate.py]: prefer this over `|=`?

**Pull request merged**